### PR TITLE
Assistant workspace access, /v1 usage stats, Overview polish

### PR DESCRIPTION
**Pull request by Th0rgal**
Five changes stacked on one branch, each verified live on localhost:3001 / prod data.

## assistant: `workspace_bash` MCP tool + exec endpoint on the mission exec layer

Paloma (the Hermes assistant gateway) runs on the host with only the gateway service env, so workspace-configured secrets (`GH_TOKEN`, SSH keys, ...) were invisible to her shell — only missions saw them.

- New `workspace_bash` tool in assistant-mcp → `POST /api/workspaces/:id/exec`, defaulting to the assistant's workspace. Env vars are read fresh from workspace config on every call; no secrets copied into the gateway environment.
- `exec_workspace_command` rewritten on `WorkspaceExec::output` — the same execution layer missions use. Merges workspace `env_vars` on both host and container paths, and reuses a running container leader via nsenter instead of failing with "Directory tree … is currently busy" (reproduced live on the assistant workspace). Timeout enforced inside the workspace via `timeout(1)`; unused `stdin` field dropped (endpoint had no consumers).

## routing: count /v1 router usage in the providers usage stats

The OpenAI-compatible `/v1` proxy only recorded token usage into the in-memory health tracker, so external clients routed through it (assistant gateways, IDEs with proxy keys) never appeared on the Providers usage page — `GET /api/ai/usage/summary` aggregates `mission_events`, and router requests have no mission.

- New `proxy_usage` table in the per-user mission store; the proxy records each request's tokens (model normalized, list-price cost estimated at write time) for both non-streaming and streaming responses (usage sink invoked from `track_stream_health` when the final SSE usage event arrives).
- `get_ai_usage_summary` merges proxy rows into `by_model` / `by_day` / `by_hour` and the totals.

## dashboard: automations panel — quiet stale signal, bulk selection, full-height sidebar

- The boxed STALE badge had its 9px text floating at the top of a 20px min-height box and doubled the amber dot's signal → replaced with a quiet lowercase "stale" in the meta line; trigger chips get explicit `h-4` + `leading-none` centering.
- Selection mode for bulk cleanup: Select → per-row checkboxes, toolbar with All / Stale quick-picks and two-step Remove (click → Confirm N). Deleting nine stale automations is three clicks instead of eighteen.
- The panel was capped at `max-h-80`, wasting the sidebar's lower half → sidebar is now a flex column and the panel stretches to the full remaining height. Title shortened to "Automations" so the header fits one line at w-72.

## control: thinking/streaming bubbles as compact pills matching tool rows

Active Thinking/Streaming groups auto-expanded into a tall content card, breaking the transcript rhythm next to the one-line tool pills (`Bash(command, description) <1s`). Collapsed by default: a 26px pill with pulsing icon + live timer, content one click away (or in the Thinking side panel). Also keeps the liveness pill visible while an active group's content is still empty (previously rendered nothing and suppressed the Agent-is-working pill, leaving a dead gap). Verified live: all transcript liveness rows measure a uniform 26px.

## dashboard: Overview stats footer 2x2 below xl

Four stat cards across truncate between the kanban and the w-72 sidebar on narrower windows → 2x2 grid below `xl`, tighter gap.

## Validation

- `cargo fmt --check` / `cargo build` / `cargo test --lib` 913/913
- dashboard: `tsc --noEmit`, `vitest run` 216/216
- Live checks: workspace exec env injection on the assistant workspace (GH_TOKEN visible), bulk-select flow (10 stale selected in one click), pill heights measured at 26px, stats grid at 1180/1500px

---

> [!NOTE]
> **Medium Risk**
> Touches mission event projection, Codex cancellation/goal loops, and large control-client transcript logic—regressions could affect history loading, goal stop, or scroll/virtualization; LLM settings migration removes browser-stored API keys for titles.
> 
> **Overview**
> Mission history can load with **`profile=conversation`**, which keeps recent tool pairs fully hydrated and collapses older completed tools into **`tool_stub`** rows (byte counts + lazy metadata). The control UI loads full payloads via **`getMissionToolCallEvents`** when a stub is expanded, and history/snapshot/delta fetches pass **`trace_tail`** so paging stays consistent.
> 
> **Control transcript UX:** thinking/stream groups stay **collapsed by default** (compact pills); tool groups **expand upward** with scroll anchoring and synchronous virtualizer re-measure to avoid flash; shared-file previews **portal to `document.body`** and use taller full-viewport readers. **Stall warnings** move from mid-list to a **`StallBar`** above the composer; **`GoalBar`** gains exit/stop for native goal loops (cancel when running, deactivate automation when idle). Client-side **auto mission titles** and the **localStorage LLM client** are removed.
> 
> **Settings:** the LLM page is reworked around **`GET /api/settings/llm-roles`** (Ask assistant model override + read-only metadata/titles resolution); Ask model config is removed from Data settings. **Hermes assistant** settings add **remote proxy** status, token rotate/apply, and copy helpers.
> 
> **Other:** Overview stats go **2×2 below `xl`**; automations sidebar gets **bulk select/remove** and full-height layout; Ask panel **folds consecutive tool steps**; global CSS adds surface sheen/canvas glow and scrollbar tweaks; **Codex** goal cancel defers to **`thread/goal/clear`** via **`ExecutionEvent::Cancelled`**; Ask LLM client parses **OpenCode-style XML `<tool_call>`** text when native tool_calls are missing.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d6432d64f1435441ba216687870ca8d09a0cb5e3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## orchestrator: Codex auth preflight mirrors launch credentials

`get_backend_auth_status` now answers whether Codex can actually launch instead of only checking the generic OpenAI API-key path. The Codex branch now counts API keys, canonical OpenAI/`codex` OAuth credentials, OpenAI OAuth rotation accounts from `ai_providers.json`, and valid runtime Codex `auth.json` files. It also inspects `auth_mode` so `chatgpt` requires access + refresh tokens and `apikey` requires a key.

Additional validation for this change:

- `cargo fmt --all --check`
- `cargo test --bin orchestrator-mcp`

### CODE REVIEW

File: dashboard/src/components/active-automations.tsx [hunk 1]
```diff
@@ -91,33 +93,150 @@ export function ActiveAutomations({ missions, runningMissionIds }: ActiveAutomat
...
+      setConfirmBulk(true);
+      return;
+    }
+    setBulkBusy(true);
+    try {
+      await Promise.all(
+        Array.from(selected).map((id) => deleteAutomation(id).catch(() => null)),
+      );
+      mutate();
+      exitSelection();
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/components/active-automations.tsx]: Fixed in 70dad81a: bulk remove now uses Promise.allSettled, keeps the failed ids selected for retry, and surfaces a toast ("Failed to remove N of M automations") instead of silently exiting selection mode. mutate() still runs so successful deletes reflect immediately.

File: dashboard/src/components/active-automations.tsx [hunk 2]
```diff
@@ -167,86 +305,127 @@ function AutomationRow({
...
           )}
           {mission && (
-            <span className="truncate">
+            <span className="flex min-w-0 items-center truncate">
+              {stale && <span className="pr-1 text-[rgb(var(--foreground)/0.25)]">·</span>}
               {mission.status}
-              <span className="px-1 text-[rgb(var(--foreground)/0.25)]">·</span>
             </span>
           )}
+          <span className="flex-shrink-0 text-[rgb(var(--foreground)/0.25)]">·</span>
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/components/active-automations.tsx]: Fixed in d074e09e: the separator before the age label now only renders when something precedes it (stale flag or mission status), so rows without a linked mission show a clean "2h ago" instead of "· 2h ago".

File: src/api/proxy.rs [hunk 3]
```diff
@@ -873,6 +920,7 @@ async fn chat_completions(
                     account_id,
                     None,
                     entry.subscription_key.clone(),
+                    Some(proxy_usage_sink(state.clone(), entry.model_id.clone())),
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/proxy.rs:923]: Fixed in bdd97bdf: the Anthropic adapter now captures input_tokens at message_start and attaches OpenAI-style usage to the finish chunk built from message_delta (cumulative output_tokens); the Google adapter folds usageMetadata (prompt + candidates + thoughts) into usage on the finish chunk. track_stream_health's scanner now sees usage on both adapter routes and proxy_usage_sink records them. Covered by two new stream tests (anthropic_stream_surfaces_usage_on_finish_chunk, google_stream_surfaces_usage_on_finish_chunk).

File: src/api/workspaces.rs [hunk 4]
```diff
@@ -1212,105 +1217,46 @@ async fn exec_workspace_command(
...
-        }
+        Err(_) => Ok(Json(ExecCommandResponse {
+            exit_code: -1,
+            stdout: String::new(),
+            stderr: format!(
+                "Command timed out after {} seconds (exec layer)",
+                timeout_secs
+            ),
+            timed_out: true,
+        })),
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/workspaces.rs:1264]: Fixed in bdd97bdf: the in-workspace timeout(1) now runs with -k 5, escalating TERM→KILL inside the workspace 5s after expiry, so TERM-ignoring commands die where the host can actually reach them (coreutils timeout signals the command's process group). The outer tokio timeout (+15s) remains as a guard for exec-layer hangs — on that path the dropped output() future kills the timeout/nsenter wrapper, and the inner -k has already reaped the command tree.

File: dashboard/src/app/control/control-client.tsx [hunk 5]
```diff
@@ -3606,10 +3617,14 @@ const ToolCallItem = memo(function ToolCallItem({
...
             )
           ) : (
+            isLazy ? (
+              item.loading ? "loading" : "hidden"
+            ) : (
             <>
               <LiveDuration startTime={item.startTime} />
               ...
             </>
+            )
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/control-client.tsx]: Fixed in 51155047: lazy tool rows now stay neutral while details are hidden/loading, and success/error icons only render after the row is hydrated with an actual result.

File: dashboard/src/app/control/control-client.tsx [hunk 6]
```diff
@@ -4849,7 +4957,10 @@ export default function ControlClient() {
 
       if (!sorted) {
         try {
-          const snapshot = await getMissionSnapshot(id);
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/control-client.tsx:4852]: Fixed in 51155047: both delta history paths now pass profile: "conversation" and traceTail, matching initial snapshot and older-history pagination so reconnect/cache merge does not rehydrate old tool payloads.

File: dashboard/src/app/control/events-reducer.ts [hunk 7]
```diff
@@ -477,6 +482,49 @@ export function eventsToItemsImpl(
...
+          resultBytes:
+            typeof event.metadata?.result_content_bytes === "number"
+              ? event.metadata.result_content_bytes
+              : undefined,
+        };
+        toolCallMap.set(toolCallId, items.length);
+        items.push(toolItem);
+        break;
+      }
+
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/events-reducer.ts:527]: Fixed in 51155047: when a full tool_result is replayed onto a lazy stub, the reducer now treats that as hydration and clears lazy/loading while marking hasResult. Added a reducer test for this case.

File: dashboard/src/app/control/control-client.tsx [hunk 8]
```diff
@@ -10877,7 +11092,7 @@ export default function ControlClient() {
             <div
               ref={containerRef}
               data-testid="chat-scroll-container"
-              className="flex-1 overflow-y-auto px-6 pt-6 pb-2"
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/control-client.tsx:10880]: Fixed in 5b268961: delta merge now treats incoming lazy tool stubs as updates for existing tool rows with the same id, carrying completion metadata/endTime into the live row instead of dropping the stub as a duplicate.

File: src/api/control.rs [hunk 9]
```diff
@@ -4792,6 +4801,111 @@ fn summarize_inactive_stream_events(events: Vec<mission_store::StoredEvent>) ->
...
+    let mut tool_tail: Vec<(String, i64)> = pairs
+        .iter()
+        .map(|(tool_call_id, pair)| (tool_call_id.clone(), pair.latest_sequence))
+        .collect();
+    tool_tail.sort_by_key(|(_, sequence)| std::cmp::Reverse(*sequence));
+    let keep_full: HashSet<String> = tool_tail
+        .into_iter()
+        .take(trace_tail)
+        .map(|(tool_call_id, _)| tool_call_id)
+        .collect();
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/control.rs]: Fixed in 5b268961: the conversation projection now uses a mission-global recent distinct tool id set from SQLite (ORDER BY MAX(sequence) DESC), so paged older responses are projected against the mission trace tail rather than the current page. Added a regression test for older-page stubbing.

File: dashboard/src/app/control/control-client.tsx [hunk 10]
```diff
@@ -9701,11 +9916,35 @@ export default function ControlClient() {
...
+                  changed = true;
+                  return {
+                    ...item,
+                    lazy: true,
+                    loading: false,
+                    hasResult: incoming.hasResult,
+                    contentBytes: incoming.contentBytes,
+                    resultBytes: incoming.resultBytes,
+                    endTime: incoming.endTime ?? item.endTime,
+                  };
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/control-client.tsx:9974]: Fixed in 6aee576b: delta lazy stubs now update only tool rows that do not already have a loaded result, so a user-hydrated row cannot be downgraded back to hidden by later catch-up metadata.

File: src/api/control.rs [hunk 11]
```diff
@@ -4792,6 +4801,100 @@ fn summarize_inactive_stream_events(events: Vec<mission_store::StoredEvent>) ->
...
+        pair.latest_sequence = pair.latest_sequence.max(event.sequence);
+        if event.event_type == "tool_call" {
+            pair.call_content_bytes = event.content.len();
+        } else {
+            pair.has_result = true;
+            pair.result_sequence = Some(event.sequence);
+            pair.result_timestamp = Some(event.timestamp.clone());
+            pair.result_content_bytes = event.content.len();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/control.rs:4845]: Fixed in 6aee576b: stubs now get has_result/result metadata from persisted per-tool summaries loaded from SQLite, so a page boundary between tool_call and tool_result still produces correct completion metadata. Added a regression test for that boundary case.

File: src/api/control.rs [hunk 12]
```diff
@@ -4840,11 +4956,38 @@ pub async fn get_mission_events(
...
+            .into_iter()
+            .collect::<Vec<_>>();
+        let tool_summaries = control
+            .mission_store
+            .get_tool_call_summaries(mission_id, &tool_call_ids)
+            .await
+            .map_err(internal_error)?;
+        summary.events =
+            project_conversation_events(summary.events, &keep_full_tool_call_ids, &tool_summaries);
+        summary.summarized_count = summary.events.len();
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply src/api/control.rs:4991]: Fixed in 020d50b0. The conversation projection now only converts an older tool_call into a lazy tool_stub when the persisted/event summary shows a result exists. In-flight tool calls without a tool_result stay fully hydrated, preserving args and active UI state even when they fall outside the recent trace tail. Added regression coverage with conversation_projection_keeps_inflight_tools_full.

File: dashboard/src/app/control/control-client.tsx [hunk 13]
```diff
@@ -11343,10 +11590,30 @@ export default function ControlClient() {
                     <GoalBar
                       objective={goal.objective}
                       statusLabel={statusLabel}
+                      running={isBusy}
+                      onExit={
+                        activeMissionId
+                          ? (running) => handleExitGoal(activeMissionId, running)
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/control-client.tsx:11578]: Fixed in 8e445811. GoalBar now receives `isTurnInFlight` (`viewingRunState !== "idle"`), so a goal exit during `waiting_for_tool` takes the Stop path and calls `cancelMission` on the blocked turn.

File: dashboard/src/app/control/control-client.tsx [hunk 14]
```diff
@@ -6853,6 +7070,56 @@ export default function ControlClient() {
...
+        setGoalInfoByMission((prev) => {
+          const next = { ...prev };
+          delete next[missionId];
+          return next;
+        });
+        toast.success(
+          running
+            ? "Goal loop stopped; current turn interrupted"
+            : "Goal loop ended",
+        );
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/control-client.tsx:7121]: Fixed in 8e445811. `handleExitGoal` now only swallows the `updateAutomation` failure while running (where the observer flips the row on the aborted GoalStatus); while idle it rethrows, so the bar stays and an error toast shows instead of a false success.

File: dashboard/src/app/control/control-client.tsx [hunk 15]
```diff
@@ -8060,40 +8335,6 @@ export default function ControlClient() {
...
-                setCurrentMission((m) =>
-                  m?.id === targetMissionId ? { ...m, title } : m,
-                );
-                setViewingMission((m) =>
-                  m?.id === targetMissionId ? { ...m, title } : m,
-                );
-              }
-            });
-          }
-        }
```
[the remark this answers is not in the document]
Comment (Th0rgal) [reply dashboard/src/app/control/control-client.tsx:8096]: Fixed in d6432d64. The live tool_result merge now clears lazy/loading alongside setting result/endTime, so a hydrated stub row immediately renders the full done state.
